runtime/util/hshelp.c

```diff
@@ -1703,72 +1703,216 @@ fixJNIRefs(J9VMThread * currentThread, J9HashTable * classPairs, BOOLEAN fastHCR
 }
 
 static jvmtiIterationControl
-fixMemberNamesObjectIteratorCallback(J9JavaVM *vm, J9MM_IterateObjectDescriptor *objectDesc, void *userData)
+prepareToFixMemberNamesObjectIteratorCallback(J9JavaVM *vm, J9MM_IterateObjectDescriptor *objectDesc, void *userData)
 {
-	J9ThreadHashPair *data = userData;
+	J9PrepareToFixMemberNameIterData *data = userData;
 	J9VMThread *currentThread = data->currentThread;
-	J9HashTable *classHashTable = data->userData;
+	J9HashTable *classHashTable = data->classHashTable;
 	j9object_t object = objectDesc->object;
 	J9Class *clazz = J9OBJECT_CLAZZ_VM(vm, object);
 
 	if (clazz == J9VMJAVALANGINVOKEMEMBERNAME_OR_NULL(vm)) {
-		U_64 vmindex = J9OBJECT_U64_LOAD(currentThread, object, vm->vmindexOffset);
-		if (0 != vmindex) {
+		UDATA vmtarget = (UDATA)J9OBJECT_U64_LOAD(currentThread, object, vm->vmtargetOffset);
+		if (0 != vmtarget) {
 			J9JVMTIClassPair exemplar;
 			J9JVMTIClassPair *result = NULL;
 
 			j9object_t membernameClazz = J9VMJAVALANGINVOKEMEMBERNAME_CLAZZ(currentThread, object);
-			exemplar.replacementClass.ramClass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, membernameClazz);
+			exemplar.originalRAMClass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, membernameClazz);
 			result = hashTableFind(classHashTable, &exemplar);
 			if (NULL != result) {
 				jint flags = J9VMJAVALANGINVOKEMEMBERNAME_FLAGS(currentThread, object);
-				if (J9_ARE_ANY_BITS_SET(flags, MN_IS_FIELD)) {
-					/* Update vmtarget to vmindex->offset */
-					J9JNIFieldID *fieldID = (J9JNIFieldID *)(UDATA)vmindex;
-					J9ROMFieldShape *romField = fieldID->field;
-					UDATA offset = fieldID->offset;
-
-					if (J9_ARE_ANY_BITS_SET(romField->modifiers, J9AccStatic)) {
-						offset |= J9_SUN_STATIC_FIELD_OFFSET_TAG;
-						if (J9_ARE_ANY_BITS_SET(romField->modifiers, J9AccFinal)) {
-							offset |= J9_SUN_FINAL_FIELD_OFFSET_TAG;
-						}
+				if (J9_ARE_ANY_BITS_SET(flags, MN_IS_METHOD | MN_IS_CONSTRUCTOR | MN_IS_FIELD)) {
+					if (J9_ARE_ANY_BITS_SET(flags, MN_IS_METHOD | MN_IS_CONSTRUCTOR)) {
+						/* Set vmindex temporarily to the J9JNIMethodID for vmtarget. The method ID will
+						 * be updated first, and then it will be used to fix the MemberName afterward.
+						 */
+						J9JNIMethodID *methodID = currentThread->javaVM->internalVMFunctions->getJNIMethodID(currentThread, (J9Method*)vmtarget);
+						J9OBJECT_U64_STORE(currentThread, object, vm->vmindexOffset, (U_64)(UDATA)methodID);
+					}
+
+					/* Temporarily take over vmtarget as the next pointer for a linked list of all affected MemberName instances. */
+					if (NULL != data->lastAffectedMemberName) {
+						J9OBJECT_U64_STORE(currentThread, data->lastAffectedMemberName, vm->vmtargetOffset, (U_64)(UDATA)object);
+					}
+
+					J9OBJECT_U64_STORE(currentThread, object, vm->vmtargetOffset, (U_64)(UDATA)NULL);
+					data->lastAffectedMemberName = object;
+					if (NULL == data->firstAffectedMemberName) {
+						data->firstAffectedMemberName = object;
```

Comment:
would this be simpler if we use
```c++
J9OBJECT_U64_STORE(currentThread, object, vm->vmtargetOffset, (U_64)(UDATA)data->firstAffectedMemberName);
data->firstAffectedMemberName = object;
```
and avoid having `lastAffectedMemberName` field

Reply:
Simplified